### PR TITLE
Implement `SageMakerProcessingOperatorAsync`

### PR DESCRIPTION
**Pull request by rajaths010494**
- Implemented `SageMakerProcessingOperatorAsync` 
- Added async hook, trigger 
- Added test case for  `SageMakerProcessingOperatorAsync` 
- Added Documentation for the Operators
- Added example dag for the operator

The task is associated with #669. We can close the task when this PR and https://github.com/astronomer/astronomer-providers/pull/710 is merged into the main.

### CODE REVIEW

File: astronomer/providers/amazon/aws/example_dags/example_sagemaker.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    ti = get_current_context()["ti"]
+    ti.xcom_push(key="bucket_name", value=bucket_name)
+    ti.xcom_push(key="raw_data_s3_key_input", value=raw_data_s3_key_input)
+    ti.xcom_push(key="ecr_repository_name", value=ecr_repository_name)
+    ti.xcom_push(key="processing_config", value=processing_config)
+
+
+with DAG(
+    dag_id="example_sagemaker_rajath",
```
Comment (phanikumv) [inline astronomer/providers/amazon/aws/example_dags/example_sagemaker.py:103]:
```suggestion
    dag_id="example_sagemaker",
```

File: astronomer/providers/amazon/aws/example_dags/example_sagemaker.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+) as dag:
+
+    setup_aws_config = BashOperator(
+        task_id="setup_aws_config",
+        bash_command=f"aws configure set aws_access_key_id {AWS_ACCESS_KEY_ID}; "
+        f"aws configure set aws_secret_access_key {AWS_SECRET_ACCESS_KEY}; "
+        f"aws configure set default.region {AWS_DEFAULT_REGION}; ",
+    )
+
+    test_setup = set_up(
```
Comment (phanikumv) [inline astronomer/providers/amazon/aws/example_dags/example_sagemaker.py:117]: why is this required?

File: astronomer/providers/amazon/aws/operators/sagemaker.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+def serialize(result: Dict[str, Any]) -> str:
+    """Serialize any objects coming from Sagemaker API response to json string"""
+    return json.loads(json.dumps(result, cls=AirflowJsonEncoder))  # type: ignore[no-any-return]
+
+
+class SageMakerProcessingOperatorAsync(SageMakerProcessingOperator):
+    """
+    SageMakerProcessingOperatorAsync is used to analyze data and evaluate machine learning
+    models on Amazon SageMake. With Processing, you can use a simplified, managed
```
Comment (phanikumv) [inline astronomer/providers/amazon/aws/operators/sagemaker.py:23]:
```suggestion
    models on Amazon SageMaker. With SageMakerProcessingOperatorAsync, you can use a simplified, managed
```

File: docs/providers/amazon/aws/operators/sagemaker.rst [hunk 4]
```diff
@@ -0,0 +1,14 @@
+Sagemaker Operator Async
+""""""""""""""""""""""""
+
+
+SageMakerProcessingOperatorAsync Starts a processing job and poll for the status asynchronously.
```
Comment (phanikumv) [inline docs/providers/amazon/aws/operators/sagemaker.rst:5]:
```suggestion
SageMakerProcessingOperatorAsync starts a processing job on AWS Sagemaker and polls for the status asynchronously.
```

File: astronomer/providers/amazon/aws/example_dags/example_sagemaker.py (hunk 2 above)
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/example_dags/example_sagemaker.py:117]: in this method, we have set up the details for sagemaker processing job config and downstream tasks to pick from xcom of this. 
Also more details would be added as part of training and transform which is in other PR needed for downstream tasks

File: astronomer/providers/amazon/aws/operators/sagemaker.py (hunk 3 above)
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/operators/sagemaker.py]: Changed

File: docs/providers/amazon/aws/operators/sagemaker.rst (hunk 4 above)
Comment (rajaths010494) [reply docs/providers/amazon/aws/operators/sagemaker.rst]: Changed

File: astronomer/providers/amazon/aws/hooks/sagemaker.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        self.s3_hook = S3HookAsync(aws_conn_id=self.aws_conn_id)
+
+    async def describe_transform_job_async(self, job_name: str) -> Dict[str, Union[str, List[str]]]:
+        """
+        Return the transform job info associated with the name
+
+        :param job_name: the name of the transform job
+        """
+        async with await self.get_client_async() as client:
+            try:
```
Comment (pankajastro) [inline astronomer/providers/amazon/aws/hooks/sagemaker.py:29]: Remove try-catch since you are catching and raising it

File: astronomer/providers/amazon/aws/hooks/sagemaker.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                raise e
+
+    async def describe_processing_job_async(self, job_name: str) -> Dict[str, Union[str, List[str]]]:
+        """
+        Return the processing job info associated with the name
+
+        :param job_name: the name of the processing job
+        """
+        async with await self.get_client_async() as client:
+            try:
```
Comment (pankajastro) [inline astronomer/providers/amazon/aws/hooks/sagemaker.py:42]: same here

File: tests/amazon/aws/hooks/test_sagemaker.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        hook = SageMakerHookAsync(aws_conn_id="test_conn_id")
+        result = await hook.describe_transform_job_async(job_name="test_job")
+        assert result == {"test": "response"}
+
+    @pytest.mark.asyncio
+    @mock.patch("astronomer.providers.amazon.aws.hooks.sagemaker.SageMakerHookAsync.get_client_async")
+    async def test_sagemaker_transform_job_exception(self, mock_client):
+        """Assert describe_transform_job method throw exception"""
+        # mocking async context function with return_value of __aenter__
```
Comment (pankajastro) [inline tests/amazon/aws/hooks/test_sagemaker.py:63]:
```suggestion
```

File: astronomer/providers/amazon/aws/operators/sagemaker.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    def execute_complete(self, context: Context, event: Any = None) -> Dict[str, Any]:
+        """
+        Callback for when the trigger fires - returns immediately.
+        Relies on trigger to throw an exception, otherwise it assumes execution was
+        successful.
+        """
+        if event and event["status"] == "success":
+            self.log.info("%s completed successfully.", self.task_id)
+            return {"Processing": serialize(event["message"])}
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/operators/sagemaker.py:90]: Response from sagemaker describe job contains `created_time` and `last_modified_time` which is of type datetime object and it is not JSON serialisable when returning to xcom, so I am serialising it.

File: astronomer/providers/amazon/aws/hooks/sagemaker.py (hunk 5 above)
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/hooks/sagemaker.py]: removed

File: astronomer/providers/amazon/aws/hooks/sagemaker.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                raise e
+
+    async def describe_training_job_async(self, job_name: str) -> Dict[str, Union[str, List[str]]]:
+        """
+        Return the training job info associated with the name
+
+        :param job_name: the name of the training job
+        """
+        async with await self.get_client_async() as client:
+            try:
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/hooks/sagemaker.py]: removed

File: astronomer/providers/amazon/aws/triggers/sagemaker.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    SagemakerProcessingTrigger is fired as deferred class with params to run the task in triggerer.
+
+    :param job_name: name of the job to check status
+    :param poll_interval:  polling period in seconds to check for the status
+    :param aws_conn_id: AWS connection ID for sagemaker
+    :param end_time: the end time in seconds. Any
+            SageMaker jobs that run longer than this will fail.
+    """
+
+    non_terminal_states = {"InProgress", "Stopping"}
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/triggers/sagemaker.py]: Done

File: tests/amazon/aws/hooks/test_sagemaker.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+from astronomer.providers.amazon.aws.hooks.sagemaker import SageMakerHookAsync
+
+
+class TestSagemakerHookAsync:
+    @pytest.mark.asyncio
+    @mock.patch("astronomer.providers.amazon.aws.hooks.sagemaker.SageMakerHookAsync.get_client_async")
+    async def test_sagemaker_processing_job(self, mock_client):
+        """Test sagemaker async hook function to describe the processing job using Aiobotocore lib"""
+        # mocking async context function with return_value of __aenter__
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply tests/amazon/aws/hooks/test_sagemaker.py]: removed

File: tests/amazon/aws/hooks/test_sagemaker.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        )
+        hook = SageMakerHookAsync(aws_conn_id="test_conn_id")
+        with pytest.raises(ClientError):
+            await hook.describe_processing_job_async(job_name="test_job")
+
+    @pytest.mark.asyncio
+    @mock.patch("astronomer.providers.amazon.aws.hooks.sagemaker.SageMakerHookAsync.get_client_async")
+    async def test_sagemaker_transform_job(self, mock_client):
+        """Test sagemaker async hook function to describe the transform job using Aiobotocore lib"""
+        # mocking async context function with return_value of __aenter__
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply tests/amazon/aws/hooks/test_sagemaker.py]: removed

File: tests/amazon/aws/hooks/test_sagemaker.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        hook = SageMakerHookAsync(aws_conn_id="test_conn_id")
+        result = await hook.describe_training_job_async(job_name="test_job")
+        assert result == {"test": "response"}
+
+    @pytest.mark.asyncio
+    @mock.patch("astronomer.providers.amazon.aws.hooks.sagemaker.SageMakerHookAsync.get_client_async")
+    async def test_sagemaker_training_job_exception(self, mock_client):
+        """Assert describe_training_job method throw exception"""
+        # mocking async context function with return_value of __aenter__
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply tests/amazon/aws/hooks/test_sagemaker.py]: removed

File: tests/amazon/aws/hooks/test_sagemaker.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        )
+        hook = SageMakerHookAsync(aws_conn_id="test_conn_id")
+        with pytest.raises(ClientError):
+            await hook.describe_transform_job_async(job_name="test_job")
+
+    @pytest.mark.asyncio
+    @mock.patch("astronomer.providers.amazon.aws.hooks.sagemaker.SageMakerHookAsync.get_client_async")
+    async def test_sagemaker_training_job(self, mock_client):
+        """Test sagemaker async hook function to describe the training job using Aiobotocore lib"""
+        # mocking async context function with return_value of __aenter__
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply tests/amazon/aws/hooks/test_sagemaker.py]: removed

File: astronomer/providers/amazon/aws/operators/sagemaker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/operators/sagemaker.py:90]: Even in the sync version they have done the same.

**Review by phanikumv — CHANGES_REQUESTED** [review]
Enable this operator on master dag as well please

**Comment by rajaths010494** [comment]
> Enable this operator on master dag as well please

Yes added in master dag via commit bc35a385d2b93d36d4ac1363e669a54373f6f2f8

File: astronomer/providers/amazon/aws/triggers/sagemaker.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+    :param job_name: name of the job to check status
+    :param poll_interval:  polling period in seconds to check for the status
+    :param aws_conn_id: AWS connection ID for sagemaker
+    :param end_time: the end time in seconds. Any
+            SageMaker jobs that run longer than this will fail.
+    """
+
+    NON_TERMINAL_STATES = ("InProgress", "Stopping")
+    TERMINAL_STATE = ["Failed"]
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/triggers/sagemaker.py]: Changed

File: astronomer/providers/amazon/aws/operators/sagemaker.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    :param aws_conn_id: The AWS connection ID to use.
+    :param wait_for_completion: Even if wait is set to False, in async we will defer and
+        the operation waits to check the status of the processing job.
+    :param print_log: if the operator should print the cloudwatch log during processing
+    :param check_interval: if wait is set to be true, this is the time interval
+        in seconds which the operator will check the status of the processing job
+    :param max_ingestion_time: The operation fails if the processing job
+        doesn't finish within max_ingestion_time seconds. If you set this parameter to None,
+        the operation does not timeout.
+    :param action_if_job_exists: Behaviour if the job name already exists. Possible options are "increment"
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/amazon/aws/operators/sagemaker.py:43]: if job name exists what action is to be taken.

File: .circleci/integration-tests/master_dag.py [hunk 17]
```diff
@@ -111,6 +111,7 @@ def prepare_dag_dependency(task_info, execution_time):
         {"redshift_data_dag": "example_async_redshift_data"},
         {"redshift_cluster_mgmt_dag": "example_async_redshift_cluster_management"},
         {"batch_dag": "example_async_batch"},
+        {"sagemaker_dag": "example_sagemaker"},
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .circleci/integration-tests/master_dag.py]: Done

File: astronomer/providers/amazon/aws/operators/sagemaker.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (pankajastro) [reply astronomer/providers/amazon/aws/operators/sagemaker.py:43]: where we are handing that?

**Review by pankajastro — APPROVED** [review]
LGTM! Thank you @rajaths010494 for addressing the comments.

**Review by phanikumv — APPROVED** [review]
